### DISCUSSION TITLE
Send post request in server component

### DISCUSSION THREAD

**Discussion by nyzd** [vercel/next.js#42979] [Help] [answered]
I wanted to create a login page where users fill data and send post requests to API, I first tried to send the request in the server component but I can't, is there any possible way to send post reqeust in the server component? btw I'm new to the next.js and SSR.

**Comment by icyJoseph** [comment, marked as answer]
Hi,

## With Pages

Using Pages you just setup a form and a fire a post request to a `api/route`.

You could also use https://github.com/smeijer/next-runtime, which is a 3rd party library that allows you to use a page with `GetServerSideProps`, and receive post requests to that endpoint directly!

## With App

Things are still on early development, but for the time being, you need a client component to gain interactivity and be able to send post requests to an API route. Server components can't have handlers, and other pieces of un-serializable data. 

Here for example, https://beta.nextjs.org/docs/data-fetching/mutating, the server component renders a bunch of client components which in turn are the ones responsible for performing mutations, that's where you'd have to make a request for an `api/route`.

Could you give a more precise example, perhaps with pseudo-code of what you are trying to achieve?

**Reply by nyzd** [reply]
So I want to handle inputs in the client component and send them to the server component and the server component sends the request to the API. perhaps I'm thinking the wrong way. Also, my API is a separate app.

**Reply by icyJoseph** [reply]
Right, it'd be really cool if we could define a by method handler on server components, because today nothing stops a POST request from getting HTML, including calls to pages btw. 



That's what `next-runtime` takes advantage of.

Same seems to apply to `appDir`, here I use middleware to log the request method:

![Screenshot 2022-11-16 at 11 53 47](https://user-images.githubusercontent.com/21013447/202161921-d25efa7e-1a82-41f2-874e-29fe474c22dc.png)

In terms of sending data to a server-component I am not sure, other than through mutations defined in the docs I shared earlier, https://beta.nextjs.org/docs/data-fetching/mutating, you could use URL queries, or cookies perhaps. You need to call the refresh method.

I think for now, send the request to the API via API routes, and then call refresh?

**Reply by nyzd** [reply]
Ok, I will try it, thank you for your help.
